add SDHC support for TI MMCSD subsystem

natto1784 opened the pull request:
This pull request does the following two things

- Adds TI AM654 SDHC driver that supports both SD cards and eMMC
- Adds SDHC nodes for AM243x EVM
- Adds SDHC nodes for AM62 SK

- [x]  ~Note that the PR will remain in draft stage until https://github.com/zephyrproject-rtos/zephyr/pull/91701 is merged. Once it is merged, will drop https://github.com/zephyrproject-rtos/zephyr/commit/b37530d82bdc8a42eddbb51a2a0fa93b9195184f~

### Device Testing

#### eMMC

eMMC has been tested for speeds upto HS200 (see Test Configuration Matrix).

#### SD card

The following devices were tested (see Test Configuration Matrix) at various speed modes:

| Device | Type | Max Speed Tested | Notes |
|--------|------|------------------|-------|
| Samsung EVO Plus | SD Card | **SDR104** (200 MHz) | Full UHS-I support verified |
| SanDisk Ultra | SD Card | **SDR50** (50 MHz) | Stable operation at SDR50 |
| Embedded eMMC | eMMC | **HS200** (200 MHz) | High-speed eMMC mode working |

#### SDIO

Tested TI Simplelink CC3351 M.2 module for basic functions
- Probing OCR
- Reading Func data from CIS
- Reading CCCR

### Test Configuration Matrix
The driver has been tested with different combinations of Auto-Stop and ADMA features using `tests/drivers/disk/disk_access`:

| Auto-Stop | ADMA | Configuration | Test Result |
|-----------|------|---------------|-------------|
| 0 | 0 | Both disabled |  Pass |
| 0 | 1 | ADMA only |  Pass |
| 1 | 0 | Auto-Stop only |  Pass |
| 1 | 1 | Both enabled |  Pass |

Note that when building the test, only one of the sdhci nodes can be present, so the user must delete one node manually, this can be added in an overlay

### Notes

The compatible string and naming scheme is borrowed from Linux to maintain DT compatibility but it gets ugly at times and also makes little sense being named after a particular SoC (AM654), so please tell if this is something that is not desired

natto1784 committed:
drivers: sdhc: add TI AM654 SDHCI driver

Add the driver for TI's interrupt based AM654 SD Host Controller that can
be used with both eMMC and SD cards.

Signed-off-by: Amneesh Singh <amneesh@ti.com>

natto1784 committed:
boards: am243x_evm/am2434/r5f0_0: add SDHC nodes

Add SD host controller nodes and also assign them interrupts and power
domains. Additionally also setup eMMC and SD card as disks.

Signed-off-by: Amneesh Singh <amneesh@ti.com>

natto1784 commented:
not stale

natto1784 committed:
boards: sk_am62/am6234/a53: add SDHC nodes

Add SD host controller nodes t and also assign them interrupts.
Additionally also setup eMMC and SD card as disks, and set up a node for
the SDIO instance.

Signed-off-by: Amneesh Singh <amneesh@ti.com>

soumya-TI commented on dts/bindings/sdhc/ti,am654-sdhci.yaml:
```diff
@@ -0,0 +1,126 @@
+    type: int
+
+  # HS200 input tap delay as calculated during tuning
+  ti,otap-del-sel-hs200:
```
nit: keep the hs200 after the sdr104, may be all otaps and itaps sequencially

soumya-TI commented on drivers/sdhc/sdhc_ti_am654.c:
```diff
@@ -0,0 +1,1977 @@
+#ifdef CONFIG_SDHC_TI_AM654_ENABLE_ADMA
+	/* ADMA descriptors */
+	struct ti_am654_adma2_descriptor descs[MAX(CONFIG_SDHC_TI_AM654_ADMA_DESC_LEN, 1)];
+#ifdef CONFIG_SDHC_HANDLE_CACHE
```
is it not implicit if we use adma, cache needs to managed. we can have these buffers declared

natto1784 commented on drivers/sdhc/sdhc_ti_am654.c:
```diff
@@ -0,0 +1,1977 @@
+}
+
+#ifndef CONFIG_SDHC_TI_AM654_ENABLE_AUTO_STOP
+static int ti_am654_request(const struct device *dev, struct sdhc_command *cmd,
```
Forward declaration for the function just below (`ti_am654_request_stop_transmission`). Shall I move this to top?

natto1784 commented on drivers/sdhc/sdhc_ti_am654.c:
```diff
@@ -0,0 +1,1977 @@
+			/* any error */
+			LOG_DBG("Command Error Status: 0x%x", events >> 16);
+
+			ti_am654_reset(dev, TI_AM654_RESET_TYPE_CMD);
```
I skipped checking since we are already returning `-EIO`, do we want to return some other code?

natto1784 committed:
boards: am62l_evm/am62l3/a53: add SDHC nodes

Add SD host controller nodes and also assign them interrupts and power
domains.

Additionally also setup eMMC and SD card as disks, and set up a node for
the SDIO instance.

Signed-off-by: Amneesh Singh <amneesh@ti.com>

natto1784 commented on drivers/sdhc/sdhc_ti_am654.c:
```diff
@@ -0,0 +1,1995 @@
+
+		} else if (events & TI_AM654_NORMAL_INTR_XFER_COMPLETE) {
+			/* xfer completed successfully */
+#ifndef CONFIG_SDHC_TI_AM654_ENABLE_AUTO_STOP
```
similarly, `ti_am654_request_stop_transmission` is not defined unless `CONFIG_SDHC_TI_AM654_ENABLE_AUTO_STOP` is not defined

natto1784 commented on drivers/sdhc/sdhc_ti_am654.c:
```diff
@@ -0,0 +1,1995 @@
+	phy_ctrl_5 = ss_regs->PHY_CTRL_5;
+
+	/* modify phy_ctrl5 */
+#ifdef CONFIG_SDHC_TI_AM654_LEGACY_PHY
```
Done. Did the same wherever else possible to avoid `ifdef` and `ifndef`, with some places being exceptions due to aforementioned reasons.

natto1784 commented:
rebased

danieldegrasse commented on drivers/sdhc/sdhc_ti_am654.c:
```diff
@@ -0,0 +1,1978 @@
+#define SDHC_CACHE_ALIGNMENT (128) /* safe default */
+#endif
+
+static const uint8_t ti_am654_tuning_blk_8_bit[] = {
```
This tuning block is standard across multiple drivers: https://github.com/zephyrproject-rtos/zephyr/blob/f39ad09bd21c5bf409e13e6420e5b396d56d2ae8/drivers/sdhc/rcar_mmc.c#L1561

We should probably pull this into a shared header (not required for this PR, more just noting it as a todo)

natto1784 replied:
the register space is the same across many drivers as well (depending on the SDHC version)

danieldegrasse replied:
Yeah, there is a shared spec for SDHC peripherals that some IP blocks implement. At some point there is likely value in creating a shared set of functions that SDHC peripherals can reuse if they implement the same spec

danieldegrasse reviewed (commented):
Can you add one of these boards to `tests/drivers/sdhc` so that the driver will be built by CI? Just need to add `-sdhc` to the board's supported tag in the twister yaml file, and create an alias `sdhc0` for the SD device to test

glneo commented on drivers/sdhc/sdhc_ti_am654.c:
```diff
@@ -0,0 +1,1977 @@
+#define TI_AM654_HOST_CONTROL1_EXT_DATA_WIDTH       BIT(5)
+#define TI_AM654_HOST_CONTROL1_HIGH_SPEED_ENA       BIT(2)
+#define TI_AM654_HOST_CONTROL1_DATA_WIDTH           BIT(1)
+#define TI_AM654_HOST_CONTROL1_DATA_WIDTH           BIT(1)
```
I still see the duplicate?

natto1784 replied:
seems like I forgot to add the change, been careless lately, fixed now, thanks.

danieldegrasse commented on drivers/sdhc/Kconfig.ti:
```diff
@@ -0,0 +1,47 @@
+	help
+	  Maximum number of ADMA descriptor lines allowed
+
+config SDHC_TI_AM654_LEGACY_PHY
```
Sorry- missed this. I do think this should be a dt property, it is describing part of the hardware (and while this rarely happens, some SOCs do have multiple revisions of an IP block)

danieldegrasse replied:
Ping @natto1784- once this is changed I'm good with the PR

natto1784 replied:
thanks for another ping, I have reverted to the old behavior, using DT property instead of this Kconfig

natto1784 commented:
Resolved conflicts

bricle commented:
@vaishnavachath Could you please take a look?

natto1784 commented:
> @vaishnavachath Could you please take a look?

Vaishnav left our org a while ago, so he might not be the best point of contact for this IMO, we might need to edit MAINTAINERS file for this.

CC: @soumya-TI @glneo @Aman-Lachhiramka-ti

danieldegrasse commented:
@glneo could you take a look at this?
